scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala

```diff
@@ -342,32 +342,159 @@ class PairSCollectionFunctions[K, V](val self: SCollection[(K, V)]) {
     implicit hash: Hash128[K],
     koder: Coder[K],
     voder: Coder[V]): Seq[(SCollection[(K, V)], SCollection[(K, V)], SCollection[(K, W)])] = {
-    val bfSettings = PairSCollectionFunctions.optimalBFSettings(thatNumKeys, fpProb)
 
-    val numKeysPerPartition = if (bfSettings.numBFs == 1) thatNumKeys.toInt else bfSettings.capacity
-    val n = bfSettings.numBFs
+    val thatBfSIs = that.optimalKeysBloomFiltersAsSideInputs(thatNumKeys, fpProb)
+    val n = thatBfSIs.size
+
     val thisParts = self.partition(n, _._1.hashCode() % n)
     val thatParts = that.partition(n, _._1.hashCode() % n)
 
-    (thisParts zip thatParts).map {
-      case (lhs, rhs) =>
-        val width = BloomFilter.optimalWidth(numKeysPerPartition, fpProb).get
-        val numHashes = BloomFilter.optimalNumHashes(numKeysPerPartition, width)
-        val rhsBf = that.keys
-          .aggregate(BloomFilterAggregator[K](numHashes, width))
-          .asSingletonSideInput(BFZero[K](BFHash[K](0, 0), 0))
+    (thisParts zip thatParts zip thatBfSIs).map {
+      case ((lhs, rhs), bfsi) =>
         val (lhsUnique, lhsOverlap) = (SideOutput[(K, V)](), SideOutput[(K, V)]())
         val partitionedLhs = lhs
-          .withSideInputs(rhsBf)
+          .withSideInputs(bfsi)
           .transformWithSideOutputs(Seq(lhsUnique, lhsOverlap)) { (e, c) =>
-            if (c(rhsBf).maybeContains(e._1)) {
+            if (c(bfsi).maybeContains(e._1)) {
               lhsOverlap
             } else {
               lhsUnique
             }
           }
         (partitionedLhs(lhsUnique), partitionedLhs(lhsOverlap), rhs)
     }
+
+  }
+
+  /**
+   * Look up values from `that` where `that` is much larger and keys from `this` wont fit in memory,
+   * and is sparse in `that`. A Bloom Filter of keys in `this` is used to filter out irrelevant keys
+   * in `that`. This is useful when searching for a limited number of values from one or more vaery
+   * large tables. Read more about Bloom Filter: [[com.twitter.algebird.BloomFilter]].
+   * @group join
+   * @param thisNumKeys estimated number of keys in `this`
+   * @param fpProb false positive probability when filtering `that`
+   */
+  def sparseLookup[A: Coder](that: SCollection[(K, A)], thisNumKeys: Long, fpProb: Double)(
+    implicit hash: Hash128[K],
+    koder: Coder[K],
+    voder: Coder[V]): SCollection[(K, (V, Iterable[A]))] = {
+    val selfBfSideInputs = self.optimalKeysBloomFiltersAsSideInputs(thisNumKeys, fpProb)
+    val n = selfBfSideInputs.size
+
+    val thisParts = self.partition(n, _._1.hashCode() % n)
+    val thatParts = that.partition(n, _._1.hashCode() % n)
+
+    SCollection.unionAll(
+      (thisParts zip selfBfSideInputs zip thatParts).map {
```

Comment:
@anish749 sorry I missed this! this kind of notation is discouraged.

Reply:
i'll keep this in mind 👍 Was away the last few days :)